core/src/main/java/org/apache/iceberg/CatalogUtil.java

```diff
@@ -215,16 +216,27 @@ public static FileIO loadFileIO(
       Configuration hadoopConf) {
     LOG.info("Loading custom FileIO implementation: {}", impl);
     DynConstructors.Ctor<FileIO> ctor;
+    boolean useNoArgConstructor = true;
     try {
       ctor = DynConstructors.builder(FileIO.class).impl(impl).buildChecked();
     } catch (NoSuchMethodException e) {
-      throw new IllegalArgumentException(String.format(
-          "Cannot initialize FileIO, missing no-arg constructor: %s", impl), e);
+      LOG.error("No-arg constructor not found for {}, try to use Hadoop Configuration constructor", impl, e);
+      try {
+        ctor = DynConstructors.builder(Catalog.class).impl(impl, Configuration.class).buildChecked();
+        useNoArgConstructor = false;
```

Comment:
Could we just invoke the constructor where we load it and skip the state variable passing through here?

Something like
```java
Try {
   ctor = ..
   fileIO = ctor.newInstance
} catch {
   ctor = ... single are ctor
   fileIo = ctor.newInstance(hadoopConf)
}
```

Reply:
One additional question, do we always want the 0 arg constructor to get priority over a 1-arg? Or should we throw an error if you provide both?

Reply:
> do we always want the 0 arg constructor to get priority over a 1-arg

I think it is better to let 0 arg constructor to get priority, so it does not block any FileIO that might have both constructors.

> To avoid try nesting we could extract that to a helper function

Yeah I tried about that but there was duplicated logic for finding ClassCastException. Using a helper function is probably a way to solve it, let me update with that, thank you!